File: workbook/ch05/simple/2/tac.py
class ASTNode:
    def __init__(self, kind, value=None, children=None):
        self.kind = kind
        self.value = value
        self.children = children if children else []

    def __repr__(self):
        return f"ASTNode(kind={self.kind}, value={self.value}, children={self.children})"


class TACGenerator:
    def __init__(self, ast):
        self.ast = ast
        self.tac = []
        self.temp_counter = 0

    def generate_tac(self, node=None):
        if node is None:
            node = self.ast

        if node.kind == "PROGRAM":
            for child in node.children:
                self.generate_tac(child)

        elif node.kind == "ASSIGN":
            var_name = node.children[0].value  # Left-hand side (variable)
            expr = node.children[1]  # Right-hand side (expression)
            result = self.generate_expression(expr)
            self.tac.append(f"{var_name} = {result}")

        elif node.kind in {"PLUS", "MINUS", "TIMES", "DIVIDE"}:
            left = self.generate_expression(node.children[0])
            right = self.generate_expression(node.children[1])
            temp_result = self.new_temp()
            op_map = {
                "PLUS": "+",
                "MINUS": "-",
                "TIMES": "*",
                "DIVIDE": "/"
            }
            self.tac.append(f"{temp_result} = {left} {op_map[node.kind]} {right}")
            return temp_result

        elif node.kind == "NUMBER":
            return node.value

        elif node.kind == "IDENTIFIER":
            return node.value

    def generate_expression(self, expr):
        if expr.kind in {"PLUS", "MINUS", "TIMES", "DIVIDE"}:
            left = self.generate_expression(expr.children[0])
            right = self.generate_expression(expr.children[1])
            temp_result = self.new_temp()
            op_map = {
                "PLUS": "+",
                "MINUS": "-",
                "TIMES": "*",
                "DIVIDE": "/"
            }
            self.tac.append(f"{temp_result} = {left} {op_map[expr.kind]} {right}")
            return temp_result

        elif expr.kind == "NUMBER":
            return expr.value

        elif expr.kind == "IDENTIFIER":
            return expr.value

    def new_temp(self):
        self.temp_counter += 1
        return f"t{self.temp_counter}"

    def get_tac(self):
        return self.tac


# example
ast = ASTNode(kind="PROGRAM", value=None, children=[
    ASTNode(kind="ASSIGN", value=None, children=[
        ASTNode(kind="IDENTIFIER", value="x", children=[]),
        ASTNode(kind="NUMBER", value=2025, children=[])
    ]),
    ASTNode(kind="ASSIGN", value=None, children=[
        ASTNode(kind="IDENTIFIER", value="y", children=[]),
        ASTNode(kind="NUMBER", value=1477, children=[])
    ]),
    ASTNode(kind="ASSIGN", value=None, children=[
        ASTNode(kind="IDENTIFIER", value="z", children=[]),
        ASTNode(kind="MINUS", value=None, children=[
            ASTNode(kind="PLUS", value=None, children=[
                ASTNode(kind="IDENTIFIER", value="x", children=[]),
                ASTNode(kind="IDENTIFIER", value="y", children=[])
            ]),
            ASTNode(kind="DIVIDE", value=None, children=[
                ASTNode(kind="TIMES", value=None, children=[
                    ASTNode(kind="NUMBER", value=5, children=[]),
                    ASTNode(kind="PLUS", value=None, children=[
                        ASTNode(kind="NUMBER", value=7, children=[]),
                        ASTNode(kind="NUMBER", value=9, children=[])
                    ])
                ]),
                ASTNode(kind="NUMBER", value=2, children=[])
            ])
        ])
    ])
])

# generate TAC from AST
tac_generator = TACGenerator(ast)
tac_generator.generate_tac()

for instruction in tac_generator.get_tac():
    print(instruction)

File: workbook/ch05/elementary/07/src/tac.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "ast.h"
#include "symbol_table.h"
#include "tac.h"


// global list of TAC instructions
TAC *tac_head = NULL;
TAC *tac_tail = NULL;

// counter for generating temporary variables
int temp_counter = 0;
int label_counter = 0;

// generate a new temporary variable
char *newTemp() {
    char *temp = malloc(16);
    snprintf(temp, 16, "t%d", temp_counter++);
    return temp;
}

char *newLabel() {
    char *label = malloc(16);
    snprintf(label, 16, "L%d", label_counter++);
    return label;
}

void emitTAC(const char *op, const char *arg1, const char *arg2, const char *result) {
    TAC *new_tac = malloc(sizeof(TAC));
    if (!new_tac) {
        fprintf(stderr, "Memory allocation failed for TAC instruction\n");
        exit(EXIT_FAILURE);
    }

    new_tac->op = strdup(op);
    new_tac->arg1 = arg1 ? strdup(arg1) : NULL;
    new_tac->arg2 = arg2 ? strdup(arg2) : NULL;
    new_tac->result = result ? strdup(result) : NULL;
    new_tac->next = NULL;

    if (!tac_head) {
        tac_head = tac_tail = new_tac;
    } else {
        tac_tail->next = new_tac;
        tac_tail = new_tac;
    }
}

void freeTAC() {
    TAC *current = tac_head;
    while (current) {
        TAC *next = current->next;
        free(current->op);
        free(current->arg1);
        free(current->arg2);
        free(current->result);
        free(current);
        current = next;
    }
    tac_head = tac_tail = NULL;
}


char *generateTAC(ASTNode *node) {
    if (!node) return NULL;

    switch (node->type) {

        case NODE_BLOCK: {
            if (strcmp(node->value, "main") == 0) {
                emitTAC("LABEL", NULL, NULL, "main"); // label: 'main'
                for (int i = 0; i < node->childCount; i++) {
                    generateTAC(node->children[i]); // assignments and CALL
                }
            } else {
                // generic block (no label)
                for (int i = 0; i < node->childCount; i++) {
                    generateTAC(node->children[i]);
                }
            }
            return NULL;
        }

        case NODE_PROC_DECL: {
            emitTAC("LABEL", NULL, NULL, node->value); // label: procedure name
            generateTAC(node->children[0]); // procedure body
            emitTAC("RETURN", NULL, NULL, NULL); // return from procedure
            return NULL;
        }

        case NODE_WHILE: {
            char *start_label = newLabel();
            char *end_label = newLabel();
            emitTAC("LABEL", NULL, NULL, start_label); // L0:

            // evaluate condition (e.g. b != 0)
            char *cond_temp = generateTAC(node->children[0]); // t2 = != t0 t1
            emitTAC("IF_NOT", cond_temp, end_label, NULL); // IF_NOT t2 GOTO L1

            generateTAC(node->children[1]); // nested block

            emitTAC("GOTO", start_label, NULL, NULL); // GOTO L0
            emitTAC("LABEL", NULL, NULL, end_label); // L1:
            return NULL;
        }

        case NODE_CONDITION: {
            if (node->childCount != 2) {
                fprintf(stderr, "Error: CONDITION node must have two children\n");
                exit(1);
            }
            if (strcmp(node->value, "#") == 0) {
                //  "#" --> "!="
                char *left = generateTAC(node->children[0]);
                char *right = generateTAC(node->children[1]);
                char *result = newTemp();
                emitTAC("!=", left, right, result);
                return result;
            }
            char *left = generateTAC(node->children[0]);
            char *right = generateTAC(node->children[1]);
            char *result = newTemp();
            emitTAC(node->value, left, right, result); // t5 = > t3 t4
            return result;
        }

        case NODE_IF: {
            char *cond_temp = generateTAC(node->children[0]); // eval condition
            char *skip_label = newLabel();
            emitTAC("IF_NOT", cond_temp, skip_label, NULL); // IF_NOT cond_temp GOTO L2

            // IF body (assignment)
            generateTAC(node->children[1]);

            emitTAC("LABEL", NULL, NULL, skip_label); // L2:
            return NULL;
        }

        case NODE_ASSIGNMENT: {
            char *temp = generateTAC(node->children[0]);
            if (temp) {
                emitTAC("=", temp, NULL, node->value); // squ = t2
            } else {
                fprintf(stderr, "Error: Assignment has no value\n");
            }
            return NULL;
        }

        case NODE_OPERATOR: {
            // operator (e.g. -, +)
            char *left = generateTAC(node->children[0]);
            char *right = generateTAC(node->children[1]);
            char *result = newTemp();
            emitTAC(node->value, left, right, result); // t8 = - t6 t7
            return result;
        }

        case NODE_EXPRESSION: {
            // expression (forward to child)
            // asymmetric tree, so we can't just return the result of the left child!
            return generateTAC(node->children[0]);
        }

        case NODE_IDENTIFIER: {
            // load variable into temporary
            char *temp = newTemp();
            emitTAC("LOAD", node->value, NULL, temp); // t0 = LOAD b
            return temp;
        }

        case NODE_NUMBER: {
            // load constant into temporary
            char *temp = newTemp();
            emitTAC("LOAD", node->value, NULL, temp); // t1 = LOAD 0
            return temp;
        }

        case NODE_CONST_DECL: {
            char *value_temp = generateTAC(node->children[0]);
            emitTAC("=", value_temp, NULL, node->value);
            return NULL;
        }

        case NODE_CALL: {
            emitTAC("CALL", node->value, NULL, NULL);
            return NULL;
        }

        default:
            // other nodes (VAR_DECL, etc.)
            for (int i = 0; i < node->childCount; i++) {
                generateTAC(node->children[i]);
            }
            return NULL;
    }
}

void printTAC() {
    TAC *current = tac_head;
    while (current) {

        if (strcmp(current->op, "LABEL") == 0) {
            printf("%s:\n", current->result);

        } else if (strcmp(current->op, "IF_NOT") == 0) {
            printf("IF_NOT %s GOTO %s\n", current->arg1, current->arg2); // conditional jumps

        } else if (strcmp(current->op, "GOTO") == 0) {
            printf("GOTO %s\n", current->arg1); // unconditional jumps

        } else if (strcmp(current->op, "CALL") == 0) {
            printf("CALL %s\n", current->arg1); // procedure calls

        } else if (strcmp(current->op, "LOAD") == 0) {
            printf("%s = LOAD %s\n", current->result, current->arg1); // loads

        } else if (strcmp(current->op, "RETURN") == 0) {
            printf("RETURN\n"); // procedure return

        } else if (strcmp(current->op, "=") == 0) {  // assignments
            printf("%s = %s\n", current->result, current->arg1);

        } else {
            // standard ops (+, -, !=)
            printf("%s = %s %s %s\n", current->result, current->op, current->arg1, current->arg2);
        }
        current = current->next;
    }
}

void exportTACFile(FILE *file) {
    TAC *current = tac_head;
    while (current) {
        fprintf(file, "TYPE: %s\n", current->op);
        fprintf(file, "ARG1: %s\n", current->arg1 ? current->arg1 : "NULL");
        fprintf(file, "ARG2: %s\n", current->arg2 ? current->arg2 : "NULL");
        fprintf(file, "RESULT: %s\n\n", current->result ? current->result : "NULL");  // Empty line as separator
        current = current->next;
    }
}

void exportTAC(const char *filename) {
    FILE *file = fopen(filename, "w");
    if (!file) {
        perror("Error opening file");
        exit(1);
    }
    exportTACFile(file);
    fclose(file);
}

// --- TAC parser ---

void parseTAC(const char *filename) {
    FILE *file = fopen(filename, "r");
    if (!file) {
        perror("Error opening file");
        return;
    }

    char line[128], op[32], arg1[32], arg2[32], result[32];

    while (fgets(line, sizeof(line), file)) {
        if (strncmp(line, "TYPE:", 5) == 0) {
            sscanf(line, "TYPE: %s", op);
        } else if (strncmp(line, "ARG1:", 5) == 0) {
            sscanf(line, "ARG1: %s", arg1);
        } else if (strncmp(line, "ARG2:", 5) == 0) {
            sscanf(line, "ARG2: %s", arg2);
        } else if (strncmp(line, "RESULT:", 7) == 0) {
            sscanf(line, "RESULT: %s", result);
            // Now we have a complete instruction
            printf("Parsed TAC - OP: %s, ARG1: %s, ARG2: %s, RESULT: %s\n",
                   op, arg1, arg2, result);
        }
    }

    fclose(file);
}
